### PR TITLE
HW TM support

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8557/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8558/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8620/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8622/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8624/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8623/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@brminich @yosefe

File: src/uct/ib/rc/accel/rc_mlx5_iface.c [hunk 1]
```diff
@@ -289,9 +289,9 @@ static void uct_rc_iface_preinit(uct_rc_mlx5_iface_common_t *iface, uct_md_h md,
                                  const uct_iface_params_t *params,
```
Comment (brminich) [inline src/uct/ib/rc/accel/rc_mlx5_iface.c]: uct_rc_iface_preinit -> uct_rc_mlx5_iface_preinit

File: test/gtest/ucp/test_ucp_tag_offload.cc [hunk 2]
```diff
@@ -380,10 +380,12 @@ UCS_TEST_P(test_ucp_tag_offload_multi, recv_from_multi)
     post_recv_and_check(e(2), 0u, tag + 1, UCP_TAG_MASK_FULL);
 }
 
+#if !HAVE_INFINIBAND_TM_TYPES_H
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_tag_offload.cc]: why is this needed? Would not it work with exp verbs?

File: src/uct/ib/configure.m4 [hunk 3]
```diff
@@ -353,12 +362,15 @@ AS_IF([test "x$with_ib" == xyes],
            ])
        AS_IF([test "x$with_ib_hw_tm" = xexp],
            [AC_DEFINE([IBV_EXP_HW_TM], 1, [IB Tag Matching support (EXP)])
```
Comment (brminich) [inline src/uct/ib/configure.m4]: is IBV_EXP_HW_TM still needed? Looks like you check particular exp symbols when needed

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 4]
```diff
@@ -144,20 +144,114 @@ void uct_rc_mlx5_iface_common_prepost_recvs(uct_rc_mlx5_iface_common_t *iface)
...
+    struct ibv_qp *qp;
+
+    ibdev = &md->dev;
+
+    port_num = ibdev->first_port;
+    port_attr = uct_ib_device_port_attr(ibdev, port_num);
+
+    memset(&qp_init_attr, 0, sizeof(qp_init_attr));
+
+    qp_init_attr.qp_type             = IBV_QPT_RC;
```
[the remark this answers is not in the document]
Comment (Artemy-Mellanox) [reply src/uct/ib/rc/accel/rc_mlx5_common.c]: Common code will use all transport configurations: PATH_MTU, MAX_RD_ATOMIC, TIMEOUTs, SL, PKEY etc, Not sure it's suits here.

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 5]
```diff
@@ -471,9 +569,40 @@ ucs_status_t uct_rc_mlx5_init_rx_tm(uct_rc_mlx5_iface_common_t *iface,
     }
 
     iface->super.rx.srq.quota = srq_init_attr->base.attr.max_wr;
+#elif HAVE_DECL_IBV_CREATE_SRQ_EX
+    srq_init_attr->attr.max_sge        = 1;
+    srq_init_attr->attr.max_wr         = ucs_max(IBV_DEVICE_MIN_UWQ_POST,
+                                                 config->super.super.rx.queue_len);
+    srq_init_attr->attr.srq_limit      = 0;
```
[the remark this answers is not in the document]
Comment (brminich) [reply src/uct/ib/rc/accel/rc_mlx5_common.c]: ok

File: src/uct/ib/rc/accel/rc_mlx5_common.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (brminich) [reply src/uct/ib/rc/accel/rc_mlx5_common.c]: I thought we could use the same. If not - then ok

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8640/ for details (Mellanox internal link).

**Comment by Artemy-Mellanox** [comment]
@yosefe

File: src/uct/ib/configure.m4 [hunk 6]
```diff
@@ -239,9 +239,17 @@ AS_IF([test "x$with_ib" == xyes],
        AC_CHECK_DECLS([IBV_LINK_LAYER_INFINIBAND,
                        IBV_LINK_LAYER_ETHERNET,
                        IBV_EVENT_GID_CHANGE,
-                       ibv_create_qp_ex],
+                       ibv_create_qp_ex,
+                       ibv_create_srq_ex],
                       [], [], [[#include <infiniband/verbs.h>]])
 
+       AC_CHECK_DECL(ibv_query_device_ex, [
+       AC_TRY_COMPILE([#include <infiniband/verbs.h>],
```
Comment (yosefe) [inline src/uct/ib/configure.m4]: why AC_CHECK_DECLS/AC_CHECK_FUNC is not enough?

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 7]
```diff
@@ -144,20 +144,114 @@ void uct_rc_mlx5_iface_common_prepost_recvs(uct_rc_mlx5_iface_common_t *iface)
...
+    qp_attr.path_mtu                 = IBV_MTU_512;
+
+    memset(&qp_attr.ah_attr, 0, sizeof(qp_attr.ah_attr));
+    qp_attr.ah_attr.port_num         = port_num;
+    qp_attr.ah_attr.dlid             = port_attr->lid;
+    qp_attr.ah_attr.is_global        = 1;
+    if (uct_ib_device_query_gid(ibdev, port_num, UCT_IB_MD_DEFAULT_GID_INDEX,
+                                &qp_attr.ah_attr.grh.dgid, &is_roce_v2) != UCS_OK) {
+        goto err_destroy_qp;
+    }
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: space line after

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 8]
```diff
@@ -144,20 +144,114 @@ void uct_rc_mlx5_iface_common_prepost_recvs(uct_rc_mlx5_iface_common_t *iface)
...
+    /* Modify to RTR */
+    qp_attr.qp_state                 = IBV_QPS_RTR;
+    qp_attr.dest_qp_num              = qp->qp_num;
+    qp_attr.path_mtu                 = IBV_MTU_512;
+
+    memset(&qp_attr.ah_attr, 0, sizeof(qp_attr.ah_attr));
+    qp_attr.ah_attr.port_num         = port_num;
+    qp_attr.ah_attr.dlid             = port_attr->lid;
+    qp_attr.ah_attr.is_global        = 1;
+    if (uct_ib_device_query_gid(ibdev, port_num, UCT_IB_MD_DEFAULT_GID_INDEX,
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: can we take it from ib_iface->gid ?

File: src/uct/ib/rc/accel/rc_mlx5_common.c [hunk 9]
```diff
@@ -144,20 +144,114 @@ void uct_rc_mlx5_iface_common_prepost_recvs(uct_rc_mlx5_iface_common_t *iface)
...
+                        IBV_QP_STATE | IBV_QP_TIMEOUT | IBV_QP_RETRY_CNT |
+                        IBV_QP_RETRY_CNT | IBV_QP_RNR_RETRY | IBV_QP_SQ_PSN |
+                        IBV_QP_MAX_QP_RD_ATOMIC);
+    if (ret) {
+        ucs_error("Failed to modify TM control QP to RTS: %m");
+        goto err_destroy_qp;
+    }
+
+    iface->tm.cmd_qp = qp;
+
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.c]: redundant space line

File: src/uct/ib/rc/accel/rc_mlx5_common.h [hunk 10]
```diff
@@ -245,6 +245,7 @@ typedef struct uct_rc_mlx5_iface_common {
     } rx;
     uct_ib_mlx5_cq_t                 cq[UCT_IB_DIR_NUM];
     struct {
+        struct ibv_qp                *cmd_qp;
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5_common.h]: maybe add comment that this is !=NULL if QP is created by UCX

File: src/uct/ib/rc/base/rc_iface.c [hunk 11]
```diff
@@ -585,7 +585,7 @@ UCS_CLASS_INIT_FUNC(uct_rc_iface_t, uct_rc_iface_ops_t *ops, uct_md_h md,
                                 &uct_rc_fc_pending_mpool_ops,
                                 "pending-fc-grants-only");
         if (status != UCS_OK) {
-            goto err_destroy_stats;
+            goto err_destroy_srq;
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_iface.c]: are tm.stats released in case of error here?

File: src/uct/ib/configure.m4 (hunk 6 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/configure.m4]: In some old MOFED (2.1) ibv_query_device_ex() has different prototype, two arguments.

Comment (yosefe) [reply src/uct/ib/configure.m4]: ok, probably need to add comment

File: src/uct/ib/rc/base/rc_iface.c (hunk 11 above)
Comment (Artemy-Mellanox) [reply src/uct/ib/rc/base/rc_iface.c]: tm.stats allocated in rc_mlx5_common constructor, later...

Comment (yosefe) [reply src/uct/ib/rc/base/rc_iface.c]: i see it's called from uct_rc_mlx5_init_rx_tm() after creating srq

**Comment by yosefe** [comment]
@Artemy-Mellanox need to fix conflict

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8662/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/8663/ for details (Mellanox internal link).
